sh_scrapy/diskusage.py

```diff
@@ -0,0 +1,99 @@
+"""
+DiskUsage extension.
+"""
+import os
+import logging
+from threading import Timer
+from subprocess import Popen, PIPE
+
+from twisted.internet import task
+
+from scrapy import signals
+from scrapy.exceptions import NotConfigured
+from scrapy.mail import MailSender
+from scrapy.utils.engine import get_engine_status
+
+
+logger = logging.getLogger(__name__)
+DEFAULT_CHECK_INTERVAL = 300.0
+
+
+def get_disk_usage(folders):
+    """ Get disk usage for current user.
+
+    :param folders: Folders to calculate disk usage for.
+    :type folders: a list of str
+
+    :return: ``(inodes_count, space_bytes)`` tuple
+    """
+    assert folders, 'There must be some folders to calculate disk usage'
+    inodes = space = 0
+    find_process = Popen(['find'] + folders + ['-user', str(os.getuid()),
+                         '-type', 'f', '-printf', '%s\n'], stdout=PIPE)
+    awk_process = Popen(['awk', "{i++;s+=$1}END{print i\" \"s}"],
+                        stdin=find_process.stdout, stdout=PIPE)
+    try:
+        find_process.stdout.close()
+        result = awk_process.communicate()[0]
+        logger.debug('Find result: %s', result)
+        if result and result.strip():
+            inodes, space = [int(val) for val in result.strip().split(' ')]
+    except Exception as exc:
+        logger.warning("Find exception: %s", exc)
+    return inodes, space
+
+
+class DiskUsage(object):
+
+    FOLDERS = ['/scrapy', '/tmp']
+
+    def __init__(self, crawler):
+        if not crawler.settings.getbool('DISKUSAGE_ENABLED'):
+            raise NotConfigured
+
+        self.crawler = crawler
+
+        self.space_limit = crawler.settings.getint('DISKUSAGE_SPACE_LIMIT_MB')*1024*1024
+        self.inodes_limit = crawler.settings.getint('DISKUSAGE_INODES_LIMIT')
+        self.check_interval = crawler.settings.getfloat(
+            'DISKUSAGE_CHECK_INTERVAL_SECONDS', DEFAULT_CHECK_INTERVAL)
+
+        crawler.signals.connect(self.engine_started, signal=signals.engine_started)
+        crawler.signals.connect(self.engine_stopped, signal=signals.engine_stopped)
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        return cls(crawler)
+
+    def engine_started(self):
+        inodes, space = get_disk_usage(self.FOLDERS)
+        self.crawler.stats.set_value('diskusage/inodes/startup', inodes)
+        self.crawler.stats.set_value('diskusage/space/startup', space)
+        self.task = task.LoopingCall(self._task_handler)
```

Comment:
this is a potentially long blocking call, it should be done in a separate thread, probably with `threads.deferToThread` inside `task.LoopingCall`.